src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java

```diff
@@ -0,0 +1,879 @@
+                                    && isAtlasEdge(target)
+                                    && target.asPolyLine().contains(shapePoint)))
+                    {
+                        final Iterable<Point> pointsAtLocation = this.rawAtlas.pointsAt(shapePoint);
+                        if (Iterables.size(pointsAtLocation) > 1)
+                        {
+                            // This shouldn't happen, so let's log this case and section at all
+                            // points. By definition, all stacked (same location) OSM nodes will
+                            // be collapsed to a single atlas point during raw atlas creation.
+                            logger.error("Detected more than one point at {}", shapePoint);
+                        }
+                        addPointToNodeList(shapePoint, nodesForEdge);
+                    }
+                }
+
+                if (line.isClosed() && nodesForEdge.size() == 0)
+                {
+                    // Handle isolated rings by creating a node at the start and the middle
+                    // shape point, effectively splitting it in half
+                    addPointToNodeList(polyLine.first(), nodesForEdge);
+                    addPointToNodeList(polyLine.get(polyLine.size() / 2), nodesForEdge);
+                }
+                else if (!line.isClosed())
+                {
+                    // For non-closed loops, the first and last point of the edge are always nodes.
+                    // We don't care about first/last points for rings because these can be
+                    // arbitrary and we only want to section at edge intersections.
+                    addPointToNodeList(polyLine.first(), nodesForEdge);
+                    addPointToNodeList(polyLine.last(), nodesForEdge);
+                }
+
+                // Record the edge and all its nodes
+                changeSet.createEdgeToNodeMapping(line.getIdentifier(), nodesForEdge);
+            }
+            else if (isAtlasArea(line))
+            {
+                changeSet.recordArea(line);
+            }
+            else if (isAtlasLine(line))
+            {
+                // No-op. If a line doesn't quality to be an edge or area, it will stay a line in
+                // the resulting atlas.
+            }
+        });
+    }
+
+    /**
+     * Determines if the given raw atlas {@link Line} qualifies to be an {@link Area} in the final
+     * atlas. An Atlas {@link Area} is defined as being closed, with more than 3 shape points and
+     * must not contain highway tags, be a ferry, pier or railway.
+     *
+     * @param line
+     *            The {@link Line} to test
+     * @return {@code true} if the given {@link Line} qualifies as an {@link Area}
+     */
+    private boolean isAtlasArea(final Line line)
+    {
+        return line.isClosed() && containsMoreThanThreeShapePoints(line)
+                && !Validators.hasValuesFor(line, HighwayTag.class) && !RouteTag.isFerry(line)
```

Comment:
It would be nice to create an areaFilter for the tag part of that statement.

Reply:
Good point, I'll add this!!